### ISSUE TITLE
Automatically set best perfomance option EVR or VMR9

### ISSUE THREAD

**Issue by xmedeko** [Sascha-L/WPF-MediaKit#31]
Following the #24 benchmarks by  @ASantos007 : EVR without DXVA takes more CPU then VMR9. (My experience confirms that.)

I need to automatically choose the best option, either EVR or VMR9 for the users HW and settings.

I think it may be done:
- Either improve EVR so it performs like VMR9 without DXVA.
- Or retect if DXVA is switched on and then switch to EVR. Also, switching DXVA just for the application (not system wide) could be handy. (Hint: a C# code example is at [Emby.Theater.Windows](https://github.com/MediaBrowser/Emby.Theater.Windows/blob/master/Emby.Theater.DirectShow/DirectShowPlayer.cs) ).

**Comment by Sascha-L** [comment]
We also should implement ILAVVideoSettings to allow setting DXVA2, None, etc. directly from code, so WPF MediaKit control could have a property where you can set which mode is prefered.

https://github.com/Sascha-L/WPF-MediaKit/issues/24#issuecomment-260139303

**Comment by glenn-slayden** [comment]
In what library is `ILAVVideoSettings` defined, is it managed or native, and are we willing to take an additional dependency if we're not already referencing it? Alternatively, I'm not against opposed to dynamically loading managed libraries (or conditionally `P/Invoking` against native components) which can be detected at runtime, but this means the WPF-MediaKit library would need logic for detecting AssemblyLoad (or native DLL load) failures.

**Comment by xmedeko** [comment]
@glenn-slayden {{ILAVVideoSettings}} is in the {{DirectShowLib-2005.dll}}.

**Comment by Sascha-L** [comment]
Didn't know that it is already implemented in DirectShowLib, but then it is very old and maybe not up to date.

We don't need any reference to an additional assembly. We just have to implement the interface in C# by ourself.

Here is an example how we can use it:

https://forum.doom9.org/showthread.php?p=1668772#post1668772

**Comment by xmedeko** [comment]
@Sascha-L What about to set VMR9 by default: `MediaUriElement.VideoRenderer = VideoRendererType.VideoMixingRenderer9` until we have code to switch on the DXVA?

**Comment by Rust76** [comment]
If I want to enable HW acceleration, is it enough to just set VideoRendere property to VideoRendererType.EnhancedVideoRenderer or do I need to do something else too?

Another question is how do I check that the library really utilizes GPU?

**Comment by xmedeko** [comment]
@Rust76 You have to manually set DXVA. I have just creates a new wiki [HW Acceleration](https://github.com/Sascha-L/WPF-MediaKit/wiki/Usage#performance-hw-acceleration). Also, check the issue #24.

I have no experience with monitoring GPU, check e.g. [this SO answer](https://stackoverflow.com/questions/1158712/monitor-the-graphics-card-usage) or [that SO answer](https://superuser.com/questions/160291/how-can-i-monitor-video-memory-usage).

**Comment by Rust76** [comment]
@xmedeko Thank  you for your answer. I was trying to find a way to programmatically set DXVA in the LAV settings but without success. Is it even possible?

**Comment by xmedeko** [comment]
Yes, it's possible, and that's the topic of this thread. Check `ILAVVideoSettings` and the Emby.Theater player ling from the first post. PR solving this for the WPF-MediaKit are welcome 😉

**Comment by Rust76** [comment]
@xmedeko I tried to find ILAVVideoSettings from DirectShowLib but I couldn't locate it. There was no such interface with GUID used in Emby.Theater COM import. Are you sure it's in there?

**Comment by xmedeko** [comment]
Check [`LavInterfaces.cs` in Emby.Theatre](https://github.com/MediaBrowser/emby-theater-windows/blob/master/Emby.Theater.DirectShow/LavInterfaces.cs).

**Comment by Rust76** [comment]
Yeah, I checked out that, but the point is that there's no interface in DirectShowLib with such Guid.

**Comment by xmedeko** [comment]
This interface is from [LAVFilters LAVVideoSettings](https://github.com/Nevcairiel/LAVFilters/blob/master/developer_info/LAVVideoSettings.h) , not the DirectShow. It's the filter feature, not the general DirectShow feature. If you use another filter, than it would not work. The WPF Media Kit uses LavFilters by default, if they are available.
